freshBootstrapTools: rename from stdenvBootstrapTools

amjoseph-nixpkgs opened the pull request:
###### Description of changes

This is a handoff of #182058

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] powerpc64le-linux
  - [x] mips64el-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

trofi committed:
freshBootstrapTools: rename from stdenvBootstrapTools

`stdenvBootstrapTools` was defined in both `nixpkgs` and `releases`
jobsets. This makes hydra view a bit confusing:

    $ git grep -F 'stdenvBootstrapTools =' | cat
    pkgs/top-level/all-packages.nix:  stdenvBootstrapTools = if stdenv.hostPlatform.isDarwin then
    pkgs/top-level/release.nix:      stdenvBootstrapTools = with lib;

The change renames jobset to be distinct. At least it improves grep
experience.

trofi committed:
pkgs/top-level/release.nix: disallow symbol clash between 'pkgs' and local jobs

This change exposes symbol override that accidentally caused job loss on hydra:

    $ nix repl ./release.nix
    error: jobs: Unexpected attribute collision between 'jobs' and 'pkgs': stdenvBootstrapTools

Added assert makes sure attribute clashes would not be re-introduced.

Commit:
release.nix: pass missing parameter

Commit 5643714dea562f0161529ab23058562afeff46d0 introduced a tiny
bug, neglecting to pass the `pkgs` parameter to `release.nix`.

This bug went unnoticed because commit
e663518d18e7dedce127edc26888fd997e6c1e73 had introduced a much larger
bug: an attribute collision resulted in `stdenvBootstrapTools` being
shadowed, and therefore never evaluated.  As a result, the missing
`pkgs` parameter did not lead to an error.

This commit passes the missing parameter so that fixing the
larger/earlier bug will not cause an eval failure.

vcunat commented:
Are you sure that the performance of `unionOfDisjoint` won't be a bother here?  This is used on jobsets with up to around 150k jobs.

amjoseph-nixpkgs commented:
> Are you sure that the performance of `unionOfDisjoint` won't be a bother here?

No; in fact, I'm sure its performance is quite awful.

I will fix that.

amjoseph-nixpkgs commented:
Alright, 32ca89a351f9d86e525d130dd879b708dc081e40 should get us back to O(1) and be as lazy as possible.

vcunat commented:
I'm afraid that there's no practical point in doing the checks, even if they were for free.
- `throw` just adds to the huge unusable list that noone reads: https://hydra.nixos.org/jobset/nixpkgs/trunk#tabs-errors
  _It would be good_ after _we cleaned it up, but given how long it's been an issue..._
- `abort` would be possible, but I'm not convinced that we want to stop building *everything* in case a collision happens

amjoseph-nixpkgs commented:
> EDIT: ah, maybe this `throw` will drop everything, but that's the same point like for `abort`. Though maybe it will be OK-ish, **if** it's cheap enough.

Won't ofborg CI also fail loudly for any PR that creates a collision?

amjoseph-nixpkgs commented:
> * just adds to the huge unusable list that noone reads

I'm really not surprised that nobody reads it; that's not a job for humans.  Hydra used to push notifications to the maintainers of the package from which they were `throw`n.  I still can't figure out why that functionality was removed.

amjoseph-nixpkgs commented:
> Won't ofborg CI also fail loudly for any PR that creates a collision?

Yep, [it does](https://gist.github.com/afe37bb32d081adbc0bb0b740528cbbe):

```
ofborg-eval — Complete, with errors
@ofborg
ofborg-eval-nixpkgs-unstable-jobset — nix-instantiate --arg nixpkgs { outPath=./.; revCount=999999; shortRev="b9909c1"; rev="b9909c13cfbe570edbe6990a64eded07d84bad7e"; } ./pkgs/t
Details
```

vcunat commented:
The Hydra notifications didn't work.  Let me post a real example:

> The status of Hydra job ‘nixos:release-18.03:nixpkgs.supercollider.x86_64-linux’ has changed from "Success" to "Dependency failed".  For details, see
>
>  https://hydra.nixos.org/build/70685138
>
> This may be due to 16697 commits by ...

And no, it was not notifying _maintainers_.  Well the above was extreme case, but normally we have evals after a hundred commit or so, so this approach just can't work with the current popularity of NixPkgs.

Note that most of the throw messages in that list are completely legitimate and not errors, so there's even no point in notifying.  (marked as broken or otherwise not supported on a given platform)

vcunat commented:
TL;DR: someone would have to implement a new notification system that behaves reasonably in current NixPkgs.

amjoseph-nixpkgs commented:
The [wiki says](https://nixos.wiki/wiki/Nixpkgs/Contributing) "Maintainers will receive emails from Hydra when package builds enters a failed state".  You're saying that was never true?

> The Hydra notifications didn't work. Let me post a real example:
> 
> > The status of Hydra job ‘nixos:release-18.03:nixpkgs.supercollider.x86_64-linux’ has changed from "Success" to "Dependency failed".  For details, see
> > https://hydra.nixos.org/build/70685138

That is exactly the notification I would want to get if I were the maintainer of `supercollider`!

> This may be due to 16697 commits by

Come on, that is not even close to the typical number of commits between consecutive Hydra evaluations.

> there's even no point in notifying. (marked as broken or otherwise not supported on a given platform)

These builds would *enter* a failed state only once: on the first build of a newly-created package, or if the `broken=true` flag were flipped on a package whose builds *weren't* failing.  I think notifying the maintainer is a good thing in these cases too!

vcunat commented:
Oh, my bad, (mostly) it really was sent just to `meta.maintainers`, at least those that I received.  Why did I think otherwise?  Some huge cases happened by mistake or misconfiguration of jobsets, which was IIRC the last drop for Eelco killing the feature on Hydra.nixos.org (in 2018).

Since then I've been watching https://hydra.nixos.org/dashboard/vcunat@gmail.com instead, though it's less comfortable.  I'm not sure if there's some better way.

vcunat commented:
I've never been maintainer of `supercollider` and still I got that e-mail.  I think hydra SW also has capability of notifying _committers_ or maybe it was notifying both at once.  There certainly are checkboxes for that in jobset config.

toonn reviewed (commented):
This looks pretty good to me. Seems like it doesn't revert the cross-build improvements.

This does not implement the loud warning when name collisions are introduced as was discussed in the previous PR, right?

Is there a way to test whether this fixes Hydra's bootstrap-tools job on a local machine (without Hydra)?

amjoseph-nixpkgs commented:
> This does not implement the loud warning when name collisions are introduced as was discussed in the previous PR, right?

It **does** implement the loud warning.  Actually it is more than a warning: it will `throw`, which will cause OfBorg to fail CI on any PR which attempts to introduce a collision:

https://github.com/NixOS/nixpkgs/pull/183072/files#diff-9d1ef0419a63492a93e8223ffbe653714958b8f791e68b72d4dbf8a26f67551aR209

> Is there a way to test whether this fixes Hydra's bootstrap-tools job on a local machine (without Hydra)?

I added instructions explaining how to do that in https://github.com/NixOS/nixpkgs/pull/181157

Commit:
lib/attrsets.nix: add unionOfDisjoint

amjoseph-nixpkgs commented:
Force-push to 6d025a6 rebases to latest staging and squashes the history to be more readable.  The overall diff created by the PR is unchanged.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/32

toonn commented:
@amjoseph-nixpkgs, I'm testing this and unfortunately it turns out this is based on a commit after #185297 and that broke `binutils` on Darwin, which is a dependency of `stdenv`. Fix in #189894 though so if you base this off of a newer staging commit it should be fixed again and we can move this forward.

toonn reviewed (commented):
So, `unionOfDisjoint` is where the loud warning comes from?
Does this mean it'll give a single warning for the first job in the intersection that Hydra tries to evaluate, or does Hydra not stop anyway?

amjoseph-nixpkgs commented:
> So, `unionOfDisjoint` is where the loud warning comes from?

Yes.

> Does this mean it'll give a single warning for the first job in the intersection that Hydra tries to evaluate, or does Hydra not stop anyway?

OfBorg will block the merge by failing CI.

But yes, if for some reason we stop using OfBorg, the second line of defense is Hydra, which will fail any build in which evaluation touches an attr in the intersection (which is supposed to be empty) of the two attribute sets.

I'm investigating your previous comment.

amjoseph-nixpkgs commented:
Okay, rebased onto 2a452a4cbebdd41d3399b64f442f282d819c6d4f.

Is there some way to rebase a github PR without doing a force-push?  Force-pushes are so error-prone.  I've started draftifying my PRs before every force-push so I don't accidentally trigger mass notifications...

Commit:
note that `unionOfDisjoint` is commutative, unlike //

toonn reviewed (commented):
Ok, so I've verified that both the Hydra job and the `freshBootstrapTools` attribute do what they say on the tin. Looking good!

One question, Hydra will be fine with the separate attributes failing IIUC. It'll just show a bunch of failed jobs, right? But how does OfBorg deal with this. Does it only display the first attribute that fails to evaluate?

amjoseph-nixpkgs commented:
> But how does OfBorg deal with this.

[Like this](https://github.com/NixOS/nixpkgs/pull/183072#issuecomment-1200407921).

I deliberately triggered the OfBorg failure there to make sure it worked and to capture an example of the error output.

toonn commented:
And what if there's multiple colliding attributes. What I'm getting at is that I want to avoid this becoming a tennis match, where you fix a collision, re-run, fix the next collision, etc.

It would be great if the failure could list all collisions at once. Can't we merge all the failing attributes into one attrset that has all of them and then throw the full list?

amjoseph-nixpkgs commented:
> It would be great if the failure could list all collisions at once.

Please read the discussion earlier in this PR; what you suggest is exactly what the original commit 90ab9cb2ef0fc434817d47fd3725e34951476b38 did.  It has overhead and laziness problems, as [discussed earlier in this PR](https://github.com/NixOS/nixpkgs/pull/183072#issuecomment-1196579433).

Commit:
unionOfDisjoint: use builtins.intersectAttrs

This brings two benefits:

1. The complete list of collisions is printed in the whenever any colliding
   attribute is accessed.

2. The sets are intersected using a C++ primitive, which runs in O(n) time
   (intersecting pre-sorted lists) with small constants rather than interpreted
   Nix code.

Thanks to @toonn for prompting this improvement.

amjoseph-nixpkgs commented:
@toonn, sorry, my [comment](https://github.com/NixOS/nixpkgs/pull/183072#issuecomment-1255771981) was not only abrupt but also wrong.  I apologize on both counts.

There *is* a way to print all collisions while still being maximally lazy and incurring only constant overhead.  And it is, in fact, very slightly more efficient than 99da193 due to the way the Nix interpreter works (the complete set of names in an attrset is evaluated strictly whenever any of them is accessed for the first time).

Please see my latest push 037cf2fad190.

The trick is to use `builtins.intersectAttrs` builtin which can intersect attrset names in O(n) time (they are stored internally as sorted lists) rather than the O(nlogn) best we could hope for using interpreted Nix code, or the O(n^2) in 90ab9cb.  `builtins.intersectAttrs` is C++ code, so the constant factors are better too.

Here's an example of what happens if you revert bf4312de3eb33c4967c32cdf123a2d97edeb1e4a and add an attribute named `toilet` to `nonPackageJobs` in `release.nix`:

```
$ nix-build pkgs/top-level/release.nix -A stdenvBootstrapTools.x86_64-linux
error: unionOfDisjoint: collision on stdenvBootstrapTools; complete list: stdenvBootstrapTools toilet
```

toonn commented:
Really cool, I'll rebuild both the Hydra job and the attribute and then cause some collisions and then this is ready to be merged IMO.

toonn reviewed (approved):
Awesome, tested and this seems to solve the problem. Let's get Hydra building bootstrap-tools again : )

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-macos-monthly/12330/33
